python/beeai_framework/agents/base.py

```diff
@@ -14,56 +14,68 @@
 
 
 from abc import ABC, abstractmethod
-from typing import Generic, TypeVar
+from typing import Any, Generic, TypeVar
 
-from pydantic import BaseModel
+from pydantic import BaseModel, InstanceOf
 
-from beeai_framework.agents.types import AgentExecutionConfig, AgentMeta
+from beeai_framework.agents.types import AgentMeta
 from beeai_framework.cancellation import AbortSignal
 from beeai_framework.context import Run, RunContext, RunContextInput, RunInstance
 from beeai_framework.emitter import Emitter
 from beeai_framework.memory import BaseMemory
-from beeai_framework.utils.models import ModelLike
+from beeai_framework.utils.models import ModelLike, to_model
+
+
+class BaseAgentRunOptions(BaseModel):
+    signal: InstanceOf[AbortSignal] | None = None
+
 
 TInput = TypeVar("TInput", bound=BaseModel)
-TOptions = TypeVar("TOptions", bound=BaseModel)
+TOptions = TypeVar("TOptions", bound=BaseAgentRunOptions)
 TOutput = TypeVar("TOutput", bound=BaseModel)
 
 
+class BaseAgentSchemaContainer(BaseModel, Generic[TInput, TOptions, TOutput]):
+    input: type[TInput]
+    output: type[TOutput]
+    options: type[TOptions]
+
+
 class BaseAgent(ABC, Generic[TInput, TOptions, TOutput]):
     is_running: bool = False
-    emitter: Emitter
+
+    def __init__(self) -> None:
+        self.emitter = self._create_emitter()
 
     def run(
-        self,
-        prompt: str | None = None,
-        execution: AgentExecutionConfig | None = None,
-        signal: AbortSignal | None = None,
+        self, input: ModelLike[TInput], options: ModelLike[TOptions] | None = None, /, **kwargs: Any
     ) -> Run[TOutput]:
         if self.is_running:
             raise RuntimeError("Agent is already running!")
 
+        # TODO: add error handling
+        input_formatted = to_model(self._schema.input, input)
+        options_formatted = to_model(self._schema.options, options or kwargs)
```

Comment:
This won't work as it does now since kwargs would contain the contents of both `input` and `options`. If we want to keep `run` flat we will need to either find a way to distinguish input from options, or we only flatten input and not options. I believe the following would do it:
```suggestion
        self, input: ModelLike[TInput], options: ModelLike[TOptions] | None = None, /, **kwargs: Any
    ) -> Run[TOutput]:
        if self.is_running:
            raise RuntimeError("Agent is already running!")

        # TODO: add error handling
        input_formatted = to_model(self._schema.input, input  or exclude_keys(kwargs, ["options"]))
        options_formatted = to_model(self._schema.options, options or kwargs["options"])
```

Reply:
I checked out your code locally to tinker a bit and it's actually even easier than what I put above. Give the flexibility of `to_model` we can actually just do the following and retain most of the current functionality. The only caveat would be we can't do `run("query string")` anymore, but change back to`run(prompt="query string")`
```suggestion
        self, input: ModelLike[TInput] | None = None, options: ModelLike[TOptions] | None = None, /, **kwargs: Any
    ) -> Run[TOutput]:
        if self.is_running:
            raise RuntimeError("Agent is already running!")

        # TODO: add error handling
        input_formatted = to_model(self._schema.input, input or kwargs)
        options_formatted = to_model(self._schema.options, options or kwargs)
```

Reply:
You are mixing two concepts - input/options.

With my approach, you must pass an input and all other arguments are interpreted as options.

Reply:
resolved by #562